api/agent/config.go

```diff
@@ -10,75 +9,99 @@ import (
 )
 
 type AgentConfig struct {
-	MinDockerVersion string        `json:"min_docker_version"`
-	FreezeIdleMsecs  time.Duration `json:"freeze_idle_msecs"`
-	EjectIdleMsecs   time.Duration `json:"eject_idle_msecs"`
-	MaxResponseSize  uint64        `json:"max_response_size"`
-	MaxLogSize       uint64        `json:"max_log_size"`
+	MinDockerVersion   string        `json:"min_docker_version"`
+	FreezeIdle         time.Duration `json:"freeze_idle_msecs"`
+	EjectIdle          time.Duration `json:"eject_idle_msecs"`
+	HotPoll            time.Duration `json:"hot_poll_msecs"`
+	HotLauncherTimeout time.Duration `json:"hot_launcher_timeout_msecs"`
+	AsyncChewPoll      time.Duration `json:"async_chew_poll_msecs"`
+	MaxResponseSize    uint64        `json:"max_response_size_bytes"`
+	MaxLogSize         uint64        `json:"max_log_size_bytes"`
+	MaxTotalCPU        uint64        `json:"max_total_cpu_mcpus"`
+	MaxTotalMemory     uint64        `json:"max_total_memory_bytes"`
 }
 
-var MaxDisabledMsecs = time.Duration(math.MaxInt64)
+const (
+	EnvFreezeIdle         = "FN_FREEZE_IDLE_MSECS"
+	EnvEjectIdle          = "FN_EJECT_IDLE_MSECS"
+	EnvHotPoll            = "FN_HOT_POLL_MSECS"
+	EnvHotLauncherTimeout = "FN_HOT_LAUNCHER_TIMEOUT_MSECS"
+	EnvAsyncChewPoll      = "FN_ASYNC_CHEW_POLL_MSECS"
+	EnvMaxResponseSize    = "FN_MAX_RESPONSE_SIZE"
+	EnvMaxLogSize         = "FN_MAX_LOG_SIZE_BYTES"
+	EnvMaxTotalCPU        = "FN_MAX_TOTAL_CPU_MCPUS"
+	EnvMaxTotalMemory     = "FN_MAX_TOTAL_MEMORY_BYTES"
+
+	MaxDisabledMsecs = time.Duration(math.MaxInt64)
+)
 
 func NewAgentConfig() (*AgentConfig, error) {
 
-	var err error
-
 	cfg := &AgentConfig{
 		MinDockerVersion: "17.06.0-ce",
 		MaxLogSize:       1 * 1024 * 1024,
 	}
 
-	cfg.FreezeIdleMsecs, err = getEnvMsecs("FN_FREEZE_IDLE_MSECS", 50*time.Millisecond)
+	var err error
+
+	err = setEnvMsecs(err, EnvFreezeIdle, &cfg.FreezeIdle, 50*time.Millisecond)
+	err = setEnvMsecs(err, EnvEjectIdle, &cfg.EjectIdle, 1000*time.Millisecond)
+	err = setEnvMsecs(err, EnvHotPoll, &cfg.HotPoll, 200*time.Millisecond)
+	err = setEnvMsecs(err, EnvHotLauncherTimeout, &cfg.HotLauncherTimeout, time.Duration(60)*time.Minute)
+	err = setEnvMsecs(err, EnvAsyncChewPoll, &cfg.AsyncChewPoll, time.Duration(60)*time.Second)
+	err = setEnvUint(err, EnvMaxResponseSize, &cfg.MaxResponseSize)
+	err = setEnvUint(err, EnvMaxLogSize, &cfg.MaxLogSize)
+	err = setEnvUint(err, EnvMaxTotalCPU, &cfg.MaxTotalCPU)
+	err = setEnvUint(err, EnvMaxTotalMemory, &cfg.MaxTotalMemory)
+
 	if err != nil {
-		return cfg, errors.New("error initializing freeze idle delay")
+		return cfg, err
 	}
 
-	if tmp := os.Getenv("FN_MAX_LOG_SIZE"); tmp != "" {
-		cfg.MaxLogSize, err = strconv.ParseUint(tmp, 10, 64)
-		if err != nil {
-			return cfg, errors.New("error initializing max log size")
-		}
+	if cfg.EjectIdle == time.Duration(0) {
+		return cfg, fmt.Errorf("error %s cannot be zero", EnvEjectIdle)
+	}
+	if cfg.MaxLogSize > math.MaxInt32 {
```

Comment:
not important but sticks out: this isn't out of the question for logs, all things considered. 2GB signed, right? async can run 60 minutes, we should just make it int64?